### PR TITLE
Fix past CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

I tried to launch Past CI (the 2nd round) after #20861, but there are some more fixes required: Past CI images don't install other dependencies, and we need more decorators to skip some tests if they are not installed.

### CODE REVIEW

File: utils/past_ci_versions.py [hunk 1]
```diff
@@ -4,6 +4,17 @@
 
 past_versions_testing = {
     "pytorch": {
+        "1.12": {
```
Comment (ydshieh) [inline utils/past_ci_versions.py:7]: For building the Past CI docker image with `torch 1.12.x`

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -1285,6 +1286,7 @@ def test_resume_training_with_randomness(self):
             self.assertAlmostEqual(b, b1, delta=1e-5)
 
     @slow
+    @require_accelerate
```
Comment (ydshieh) [inline tests/trainer/test_trainer.py:1289]: This test requires `accelerate`, which is installed in daily CI image, but not past CI images.

(So far the past CI avoids installing many other dependencies, but this could be changed in the future)

File: tests/pipelines/test_pipelines_object_detection.py [hunk 3]
```diff
@@ -245,6 +253,7 @@ def test_threshold(self):
         )
 
     @require_torch
+    @require_pytesseract
```
Comment (ydshieh) [inline tests/pipelines/test_pipelines_object_detection.py:256]: (So far) Past CI images don't install `pytesseract`

File: tests/models/layoutlmv2/test_tokenization_layoutlmv2.py [hunk 4]
```diff
@@ -1264,6 +1271,7 @@ def test_offsets_mapping(self):
                 self.assertEqual(sum(tokens_with_offsets["special_tokens_mask"]), added_tokens)
 
     @require_torch
+    @require_detectron2
```
Comment (ydshieh) [inline tests/models/layoutlmv2/test_tokenization_layoutlmv2.py:1274]: (So far) Past CI images don't install `detectron2`

File: tests/models/bert/test_tokenization_bert_tf.py [hunk 5]
```diff
@@ -4,15 +4,15 @@
...
 
 
-if is_tensorflow_text_available():
-    from transformers.models.bert import TFBertTokenizer
-
 if is_tf_available():
     import tensorflow as tf
 
+    if is_tensorflow_text_available():
+        from transformers.models.bert import TFBertTokenizer
```
Comment (ydshieh) [inline tests/models/bert/test_tokenization_bert_tf.py:14]: This import will trigger
```
import tensorflow as tf
```
in `src/transformers/models/bert/tokenization_bert_tf.py`

File: tests/models/gpt2/test_tokenization_gpt2_tf.py [hunk 6]
```diff
@@ -4,15 +4,15 @@
...
+from transformers.testing_utils import require_keras_nlp, require_tf, slow
 
 
-if is_keras_nlp_available():
-    from transformers.models.gpt2 import TFGPT2Tokenizer
-
 if is_tf_available():
     import tensorflow as tf
 
+    if is_keras_nlp_available():
```
Comment (ydshieh) [inline tests/models/gpt2/test_tokenization_gpt2_tf.py:13]: same here

File: src/transformers/__init__.py [hunk 7]
```diff
@@ -710,7 +710,7 @@
 
 # Tensorflow-text-specific objects
 try:
-    if not is_tensorflow_text_available():
+    if not (is_tensorflow_text_available() and is_tf_available()):
```
Comment (ydshieh) [inline src/transformers/__init__.py:713]: I don't like this change much, but the import of `TFBertTokenizer` and `TFGPT2Tokenizer` also requires `tensorflow` too.

cc @Rocketknight1

**Review by sgugger — COMMENTED** [review]
Thanks for fixing those. I'm okay with most changes except for the new checks (the existing one should check on the others basically)

File: src/transformers/__init__.py [hunk 8]
```diff
@@ -723,7 +723,7 @@
 
 # keras-nlp-specific objects
 try:
-    if not is_keras_nlp_available():
+    if not (is_keras_nlp_available() and is_tensorflow_text_available() and is_tf_available()):
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/__init__.py:726]: OK for me. I had some doubts as not sure if I should mix more checks in a single method/decorator.

**Review by sgugger — APPROVED** [review]

File: tests/models/bert/test_tokenization_bert_tf.py [hunk 9]
```diff
@@ -4,15 +4,15 @@
...
+from transformers.testing_utils import require_tensorflow_text, require_tf, slow
 
 
-if is_tensorflow_text_available():
-    from transformers.models.bert import TFBertTokenizer
-
 if is_tf_available():
     import tensorflow as tf
 
+    if is_tensorflow_text_available():
```
Comment (ydshieh) [inline tests/models/bert/test_tokenization_bert_tf.py:13]: @sgugger As `is_tensorflow_text_available` already contains the check for TF, should I revert the change here? Or it's fine to keep it this way?

File: tests/models/bert/test_tokenization_bert_tf.py [hunk 10]
```diff
@@ -32,6 +32,7 @@ def call(self, inputs):
             return out["pooler_output"]
 
 
+@require_tf
```
Comment (ydshieh) [inline tests/models/bert/test_tokenization_bert_tf.py:35]: same question - we don't need this anymore, and can keep it is as in current `main`.

**Review by LysandreJik — APPROVED** [review]
Great, thank you @ydshieh!

File: tests/models/bert/test_tokenization_bert_tf.py (hunk 9 above)
Comment (ydshieh) [reply tests/models/bert/test_tokenization_bert_tf.py:13]: - In current `main` branch, `is_tensorflow_text_available` condition is not inside `is_tf_available`. In past CI image build dockerfile, we install `.[dev]` then uninstall `tensorflow`, so `tensorflow_text` is there but `tensorflow` is removed. And this causes some tensorflow_text-related tests failed (`TFBertTokenizer` file will import `tensorflow`)

- In the 1st version of the PR, I keep `is_tensorflow_text_available` as it is, and use the combination of `is_tensorflow_text_available() and is_tensorflow_available()`. **The change at these 2 lines was necessary at that time**.

- @sgugger said we should instead change the definition of `is_tensorflow_text_available` to include `is_tensorflow_available` directly. After applying that suggestion, **we no longer need to change the 2 lines here**.

Comment (sgugger) [reply tests/models/bert/test_tokenization_bert_tf.py:13]: You can unindent this block if you want but honestly we don't really care. The tensorflow import needs to stay as it's used in the file.
